src/ghci.ts

```diff
@@ -58,17 +67,86 @@ export class Ghci implements IGhci {
         return this.ghciProcess;
     }
 
-    private write(command: string) {
+    private async write(command: string) {
         try {
-            let ghciProcess = this.getGhciProcess();
+            const ghciProcess = await this.getGhciProcess();
             ghciProcess.stdin.write(command);
         } catch (e) {
-            this.logger.error(`Failed to get GHCi process: ${e}`);
+            this.logger.error(`${e}`);
             return;
         }
     }
 
-    public writeLn(command: string) {
-        this.write(`${command}${EOL}`);
+    public async writeLn(command: string) {
+        await this.write(`${command}${EOL}`);
+    }
+
+    private async resolveGhciPath(userPath: string) {
+        /*
+        1. Try to use user's configured path
+            - expand ~ to home directory if present
+        2. If that doesn't exist, check for user's configured path on PATH
+        3. If it doesn't exist, look for ghci on PATH
+        4. If it doesn't exist, look for ghci in other known locations:
+            - $HOME/.ghcup/bin/ghci
+        5. If that doesn't exist, throw error.
+        */
+
+        const triedPaths = [];
+
+        // 1. try to use user's configured path, if it exists
+        if (userPath && userPath !== ghciCommandName) {
+            const fullGhciPath = userPath.startsWith('~')
+                ? `${homedir()}${userPath}`
+                : userPath;
+
+            if (fs.existsSync(fullGhciPath)) {
+                return fullGhciPath;
+            }
+        }
+
+        // 2. look for user's configured value on PATH
+        if (await commandExists(userPath)) {
+            return userPath;
+        }
+
+        this.showWarning(`Configured GHCI path was not found: ${userPath}`);
+        triedPaths.push(userPath);
+
+        // 3. look for GHCI on PATH
+        if (
+            userPath !== ghciCommandName &&
+            (await commandExists(ghciCommandName))
+        ) {
+            return ghciCommandName;
+        }
+
+        triedPaths.push(ghciCommandName);
+
+        // 4. try a well-known path: /usr/local/bin/ghcixxx
+        const wellKnownPath1 = `/usr/local/bin/ghci`;
```

Comment:
Instead of `wellKnownPath1` and `wellKnownPath2` it would be better to support a dynamic array of possible paths, but I think this is OK for now.